### PR TITLE
[move] `Loc::file` is a `move_symbol_pool::Symbol`

### PR DESCRIPTION
**Pull request by modocache**
Use the newly added `move_symbol_pool` module from #8782 to represent file names in Move locations. This allows us to eliminate the use of "leak string" functions, and re-enable libc leak detection in LeakSanitizer.

### CODE REVIEW

File: language/move-ir/types/src/location.rs
Comment (tnowacki) [inline language/move-ir/types/src/location.rs:18]: Probably want this in another Rust module

File: language/compiler/ir-to-bytecode/syntax/src/lexer.rs
Comment (tnowacki) [inline language/compiler/ir-to-bytecode/syntax/src/lexer.rs:126]: I would just have this be a `Symbol` or `Atom` or whatever we end up doing

File: language/move-ir/types/src/location.rs
Comment (modocache) [reply language/move-ir/types/src/location.rs]: Like we discussed in chat, I'll probably stack these commits such that the symbol table is added first. But, for now as part of this WIP, I added a commit on top of the stack that adds a new Rust module, `symbol-table`.

File: language/compiler/ir-to-bytecode/syntax/src/lexer.rs
Comment (modocache) [reply language/compiler/ir-to-bytecode/syntax/src/lexer.rs:126]: Done in the latest commit!

File: language/move-ir/types/src/location.rs
Comment (modocache) [inline language/move-ir/types/src/location.rs:25]: @tnowacki As we discussed, I'll move this `Symbol::from` conversion out of `Loc` itself and into the higher-level modules, like the source language and IR compilers.

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A symbol-pool 0.1.0 (workspace, path 'language/symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A symbol-pool 0.1.0 (workspace, path 'language/symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A symbol-pool 0.1.0 (workspace, path 'language/symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A symbol-pool 0.1.0 (workspace, path 'language/symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A symbol-pool 0.1.0 (workspace, path 'language/symbol-pool')
    * features: default
```

File: language/symbol-pool/Cargo.toml
Comment (tnowacki) [inline language/symbol-pool/Cargo.toml:2]: I'd name the crate `move-symbol-pool` ... maybe
My main concern is not making it clear it is designed for compilers (or other short lived programs) with its lack of ref counting and general leakyness

File: language/symbol-pool/src/lib.rs
Comment (tnowacki) [inline language/symbol-pool/src/lib.rs:12]: `Box<str>` at the very least here, but any reason to not just go ham with a larger fork of `string_cache`?

I'd probably just start by more or less copy-pasting the `string_cache` code and then start ripping out all of the ref-counting related items

Comment (tnowacki) [reply language/symbol-pool/src/lib.rs:12]: Then again, maybe we can:
- Start here
- Move everything in move-lang to using this (basically making `Name` a `Spanned<symbol>` instead of `Spanned<string>` 
- Then actually verify perf changes?

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

File: language/symbol-pool/src/lib.rs
Comment (modocache) [reply language/symbol-pool/src/lib.rs]: I think starting small would be my personal preference -- definitely defer to your judgement here if you're saying it's guaranteed we'll need more of the optimizations from string_cache, but my take is that we can always plug those bits in later. And, like you mentioned, verifying the perf improvements might be easier that way (I guess we could add it all now, then comment out the inlining parts later, but that seems more complex).

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

File: language/symbol-pool/src/lib.rs
Comment (tnowacki) [reply language/symbol-pool/src/lib.rs]: I think the optimizations are rather easy to add, and it will be fun to actually see what the affect is. So I'm good with this!

File: language/compiler/ir-to-bytecode/syntax/src/lexer.rs
Comment (tnowacki) [reply language/compiler/ir-to-bytecode/syntax/src/lexer.rs:127]: Looks like it is still a `&'input str` here. Am I looking at the wrong commit?

File: language/move-lang/src/errors/mod.rs
Comment (tnowacki) [inline language/move-lang/src/errors/mod.rs:37]: Shouldn't this still be a symbol?

File: language/compiler/ir-to-bytecode/syntax/src/lexer.rs
Comment (tnowacki) [reply language/compiler/ir-to-bytecode/syntax/src/lexer.rs:127]: Or do you mean on a local commit?

File: language/move-ir/types/src/location.rs
Comment (tnowacki) [reply language/move-ir/types/src/location.rs:19]: Needs to be removed after the switch :)

File: language/move-lang/src/parser/mod.rs
Comment (tnowacki) [inline language/move-lang/src/parser/mod.rs:38]: Yeah... let's add `From<String>` for `Symbol` before landing this. Finding all of these spots is going to be painful

**Review by tnowacki — CHANGES_REQUESTED** [review]
Requesting changes almost entirely for `From<String>` being missing for `Symbol`

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

**Comment by modocache** [comment]
Phew, sorry this ended up taking a little while @tnowacki, but I think this and the underlying #8782 are ready for review (and re-review) ♻️

File: language/move-ir/types/src/location.rs
Comment (tnowacki) [inline language/move-ir/types/src/location.rs:42]: Do we need the ref here? not sure what is more efficient or not 
Maybe we can mark some of these  as always inline?

**Review by tnowacki — COMMENTED** [review]
Loc stuff looks good! just have to resolve the table stuff and this should be good to go!

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-symbol-pool 0.1.0 (workspace, path 'language/move-symbol-pool')
    * features: default
```

File: language/move-ir/types/src/location.rs
Comment (modocache) [reply language/move-ir/types/src/location.rs]: Ah, no, it's not needed. I also don't know which is more efficient, but I'll remove it for now since it's an unnecessary part of this PR.

> Maybe we can mark some of these as always inline?

I'll try profiling `self` vs `&self` and marking this and the other `Loc` methods as inline! In my experience inlining isn't always beneficial as it can increase binary size and result in slower programs. But if these are on especially hot paths and the compiler isn't choosing to inline them for some reason, it could result in some nice speedups. I'll try some of the tips you wrote down the other day.

File: language/move-lang/src/parser/mod.rs
Comment (tnowacki) [inline language/move-lang/src/parser/mod.rs:115]: The prophecy fulfilled!

Comment (tnowacki) [reply language/move-lang/src/parser/mod.rs:115]: I wonder how old this TODO is lol, its probably at least from Summer 2019 if not earlier...

**Review by tnowacki — DISMISSED** [review]
Thrilled to see this go in!

File: language/move-lang/src/parser/mod.rs
Comment (modocache) [reply language/move-lang/src/parser/mod.rs:115]: First added in 837d86fb74fdc97854de4b64f4bd9ad9d219163a, from January 28 2020! 😮

**Comment by modocache** [comment]
I wonder if this may have broken source maps, since `Loc::file` is now serialized as an address value of type `NonZeroU64`. When deserialized in a separate process, that address will no longer be valid. I believe I will need to implement custom serde serialization/deserialization for `Symbol`. I'm working on a pull request for that now. (Also I'll try to add some tests that would catch something like this!)

**Pull request merged**
